# Projet 4 Auditez un environnement de données

SuperSmartMarket est une chaîne de supermarchés en plein développement en France. Cette entreprise redéfinit le concept du supermarché en intégrant des technologies de pointe pour rendre les courses rapides et agréables aux consommateurs. 

 

L’entreprise souhaite se démarquer en mettant l’accent sur l’hyper personnalisation de l’expérience en magasin avec des coupons promotionnels personnalisés, des recommandations basées sur les historiques d’achat, des recettes personnalisées en fonction des achats etc.

 

SuperSmartMarket a besoin d'agrandir son équipe Data en recrutant un Data Engineer pour l’aider à comprendre et analyser les différents flux de données de l’entreprise. Jusqu’à présent les Data Analysts se chargeaient de cette activité mais les besoins récents en data ont conduit l’entreprise à  vous recruter. 

 

Vous êtes intégré à l’équipe Data Support. Cette équipe doit garantir la sécurité et la qualité des données pour l’ensemble de l’entreprise. Elle utilise une planification de processus avec des solutions Microsoft, un entrepôt de données type OLAP et le support de l’application Microsoft PowerBI.

 

Vous recevez un mail d’Hugo, c’est le responsable de tout le pôle Business Intelligence.

## Imports utiles

In [1]:
import pandas as pd
import numpy as np
import os
from unidecode import unidecode
from datetime import datetime
from pathlib import Path
from sqlalchemy import create_engine, text
from dotenv import load_dotenv
from urllib.parse import quote_plus

## Partie 1 - Auditez l'architecture et les données de l'entreprise

De : Hugo

À : Moi

Objet : Audit de la base de données OLAP

Salut !


Je te souhaite la bienvenue dans notre équipe Data, nous sommes tous enthousiastes à l’idée de travailler avec toi.


Comme tu le sais, nous souhaitons nous démarquer par notre connaissance des clients. Mais pour le moment, nous parvenons à peine à avoir un chiffre d’affaires juste !


C’est sur ce point que je sollicite ton aide. Nous avons des incohérences dans les données que nous utilisons. Je pense que cela ne te prendra que quelques jours pour comprendre et corriger les incohérences alors que nous, nous risquerions d’y passer des semaines !


Je t’explique le problème… nos données de chiffres d’affaires ne sont pas stables dans le temps. C'est-à-dire que l’historique de chiffre d’affaires a tendance à évoluer à la hausse ou à la baisse. C’est étrange et nous ne comprenons pas ce qui se passe. 


Je te donne un exemple : le chiffre d’affaires du lundi 14 août était de 275 186,59 €. Le lendemain, c’était le 15 août et tous nos magasins étaient fermés. Pourtant, quand nous sommes retournés sur PowerBI le 16 Août, le chiffre d’affaires du 14 août était de 284 243,88 €.


Pour te permettre de commencer ton investigation, tu peux consulter notre architecture qui est en pièce jointe. Nous n’avons malheureusement jamais documenté notre base de données.


Pourrais-tu préparer : 

- un dictionnaire des données ;
- un schéma relationnel sur la base du fichier à plat de la base de données OLAP ? 

Dans un deuxième temps, pour approfondir ton investigation, peux-tu me confirmer les chiffres que je vois dans PowerBI dans ton prototype de base de données, à savoir : 

- le chiffre d’affaires total pour le 14 août (275 186,59€ ou 284 243,88€) ; 
- le chiffre d’affaires par client pour le top 10 des clients ;
- la part de chiffre d’affaires encaissé par employé.

Je te conseille de faire cela en mode POC (Proof Of Concept) : tu fais un prototype de base de données en local (SQLite, PostGré ou encore MySQL) et tu fais des requêtes SQL dans ta base de données pour vérifier et identifier les bons chiffres. 


Pour finir, peux-tu préparer un support de présentation présentant : 

- ta compréhension de l’architecture de l’entreprise ;
- le schéma relationnel que tu as créé pour ta base de données ;
- le dictionnaire des données ;
- les résultats des requêtes ;
- ta compréhension de notre problème.

Tu peux utiliser et adapter la trame en pièce jointe. Tu verras que j'ai ajouté des slides pour des recommandations futures mais on en est pas là! Concentre-toi d'abord sur ces éléments.


Merci beaucoup et bon courage !


Hugo

Responsable de l’équipe BI

### Etape 1 - Consolidez les notions importantes avant de démarrer cette mission

#### 1.1. Entrepôt de données

Un entrepôt de données (ou data warehouse) est une base de données centralisée qui collecte, organise et stocke des informations issues de diverses sources opérationnelles d'une entreprise, dans le but d'aider à la prise de décision.
Deux grandes architectures s'affrontent : celle de Bill Inmon, qui conçoit l'entrepôt comme un référentiel global et centralisé, et celle de Ralph Kimball, qui le voit comme une agrégation progressive de datamarts (bases spécialisées par domaine métier). En pratique, les deux approches sont souvent combinées.
Son fonctionnement repose sur trois principes clés :

- Intégration : les données, hétérogènes et issues de multiples sources, sont harmonisées et normalisées.

- Historisation : les données sont stables et non modifiables une fois intégrées, ce qui permet de suivre leur évolution dans le temps.

- Organisation métier : les données sont croisées transversalement selon les besoins des différents services, plutôt que cloisonnées par application.

En amont, un processus ETL (Extract, Transform, Load) filtre et prépare les données. En aval, des outils d'analyse (reporting, cubes OLAP, data mining) permettent de les exploiter.

Historiquement, le concept émerge dans les années 1960-1980, se formalise à la fin des années 1980 chez IBM, et connaît un essor majeur dans les années 1990 avec les publications de Inmon et Kimball. Depuis les années 2000, le marché explose, porté notamment par le cloud (AWS) et plus récemment par l'intelligence artificielle.

Un **entrepôt OLAP (Online Analytical Processing)** est un système de stockage et d'analyse de données multidimensionnelles, conçu pour répondre rapidement à des requêtes analytiques complexes.

**Principe**<br>
Contrairement aux bases de données transactionnelles classiques (OLTP), qui sont optimisées pour les opérations courantes (insertions, mises à jour), l'OLAP est pensé pour l'analyse et l'exploration de grandes quantités de données historiques.<br>
Les données y sont organisées sous forme de cubes multidimensionnels, où chaque dimension représente un axe d'analyse (temps, géographie, produit…) et chaque cellule du cube contient une mesure (chiffre d'affaires, quantité vendue…).

**Opérations typiques**<br>
- Drill-down : aller du général au détail (ex. : des ventes annuelles aux ventes mensuelles).
- Roll-up : agréger les données (ex. : des ventes par ville aux ventes par région).
- Slice & dice : filtrer et croiser les dimensions pour isoler un sous-ensemble de données.
Pivot : réorganiser les axes d'analyse.

**Lien avec l'entrepôt de données**<br>
L'OLAP se place en aval de l'entrepôt de données. Les données, une fois intégrées et historisées dans l'entrepôt, sont chargées dans des cubes OLAP pour être restituées aux utilisateurs métier via des outils de reporting ou de visualisation.

sources: https://fr.wikipedia.org/wiki/Entrep%C3%B4t_de_donn%C3%A9es
https://fr.wikipedia.org/wiki/Traitement_analytique_en_ligne

#### 1.2. Modèle de données en étoile

La modélisation des données est un concept fondamental de la Business Intelligence. Elle repose sur la distinction entre deux types de tables :

- Les tables de faits, qui enregistrent des observations ou événements (commandes, taux de change, températures…).
- Les tables de dimensions, qui décrivent les entités analysées (produits, personnes, lieux, temps…) et servent au filtrage et au regroupement des données.

Le modèle en étoile consiste à relier plusieurs tables de dimensions à une table de faits centrale. Son principal avantage est la flexibilité : plutôt que de modifier une table de faits existante, on enrichit le modèle en ajoutant simplement de nouvelles dimensions (canal marketing, conseiller commercial, etc.), offrant ainsi de nouveaux axes d'analyse.

Ce modèle présente également des avantages pratiques concrets :

- Économie de stockage, grâce à la non-redondance des données.
- Maintenance simplifiée : une modification dans une table de dimension (par exemple, le changement de nom d'une agence) n'impacte qu'une seule cellule, au lieu de devoir mettre à jour l'ensemble des enregistrements.

En résumé, le modèle en étoile permet de construire des analyses riches et évolutives, tout en gardant une structure claire et performante.

### Etape 2 - Prenez de la hauteur sur l'architecture et la base de données

#### 2.1. Compréhension de l'architecture et des flux de données de l'entreprise

L'architecture montre 3 étapes de circulation des données dans le système informatique de l'entreprise:

1. **La source qui correspond à la base de données en live (SQL Server)**. Les données sont générées et stockées en temps réel. Ex: transactions, ventes, clients etc. C'est le coeur du système
2. A partir de cette source, on part:<br>
   Vers le **Cube OLAP (Azure Analysis Services)** &rarr; les données sont extraites et organisées pour l'analyse. Le cube permet de les croiser selon plusieurs dimensions (temps, produit, rayon…) afin de produire des rapports et des statistiques.<br>
   Vers **le site internet et l'ERP** &rarr; les mêmes données alimentent les outils opérationnels du quotidien (le site web de l'entreprise et son logiciel de gestion interne).
3. **La restitution avec le serveur Dataviz (Power BI)**: Le cube OLAP alimente enfin Power BI, qui permet aux utilisateurs de visualiser les données sous forme de tableaux de bord et de graphiques interactifs pour aider à la prise de décision

#### 2.2. Compréhension des données présentes dans le fichier de données

In [2]:
chemin = Path(os.getcwd())

In [3]:
chemin

WindowsPath('C:/Users/sebas/Documents/cours Openclassrooms/Data Engineer/projet 4')

In [4]:
os.listdir(chemin)

['.env',
 '.git',
 '.ipynb_checkpoints',
 'Donnees',
 'presentation_sebastien_ulesie.pptx',
 'Projet4-Sebastien.ipynb',
 'Projet4_dico_donnees_Sebastien_Ulesie.xlsx',
 "Rapport+d'audit-Sebastien_ULESIE.docx",
 "Rapport+d'audit.docx",
 'Schéma+architecture+.jpg',
 'Script-smartmarket.sql',
 'SQL_architecture_projet4.architect',
 'SQL_architecture_projet4.architect~',
 'SQL_architecture_projet4.pdf',
 'SQL_architecture_projet4.png',
 'SQL_script.sql',
 'Template+de+support+de+présentation+.pptx',
 "~$pport+d'audit-Sebastien_ULESIE.docx",
 '~$presentation_sebastien_ulesie.pptx',
 '~WRL0005.tmp']

In [5]:
ventes = pd.read_excel(chemin/'Donnees/Copie+de+Extraction+cube+OLAP+-+14+aout+2024.xlsx',sheet_name=0)

In [6]:
ventes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41377 entries, 0 to 41376
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   ID_BDD       41377 non-null  object
 1   CUSTOMER_ID  41377 non-null  object
 2   id_employe   41377 non-null  object
 3   EAN          41377 non-null  int64 
 4   Date achat   41377 non-null  int64 
 5   ID ticket    41377 non-null  object
dtypes: int64(2), object(4)
memory usage: 1.9+ MB


ID_BDD : identifiant d'une vente dans la base de données

CUSTOMER ID: identifiant du client qui réalise l'achat

id_employe: identifiant de l'employé (caissier) qui enregistre l'achat

EAN: European Article Numbering &rarr; code barre du produit

Date achat: date d'achat du produit

ID ticket: numéro du ticket

In [7]:
produits = pd.read_excel(chemin/'Donnees/Copie+de+Extraction+cube+OLAP+-+14+aout+2024.xlsx',sheet_name=1)

In [8]:
produits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18040 entries, 0 to 18039
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   EAN              18040 non-null  int64  
 1   categorie        18040 non-null  object 
 2   Rayon            18040 non-null  object 
 3   Libelle_produit  18040 non-null  object 
 4   prix             18040 non-null  float64
dtypes: float64(1), int64(1), object(3)
memory usage: 704.8+ KB


EAN : code barre du produit

categorie: categorie du produit (Boissons, conserves, boulangerie etc)

Rayon: rayon dans lequel se situe le produit

Libelle_produit: nom du produit

prix: prix de vente du produit

In [9]:
clients = pd.read_excel(chemin/'Donnees/Copie+de+Extraction+cube+OLAP+-+14+aout+2024.xlsx',sheet_name=2)

In [10]:
clients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2297 entries, 0 to 2296
Data columns (total 2 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   CUSTOMER_ID       2297 non-null   object        
 1   date_inscription  2297 non-null   datetime64[ns]
dtypes: datetime64[ns](1), object(1)
memory usage: 36.0+ KB


CUSTOMER ID: identifiant du client 

date_inscription: date d'inscription du client dans la base, certainement via un programme de fidélité

In [11]:
calendrier = pd.read_excel(chemin/'Donnees/Copie+de+Extraction+cube+OLAP+-+14+aout+2024.xlsx',sheet_name=3)

In [12]:
calendrier.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1999 entries, 0 to 1998
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   date          1999 non-null   int64         
 1   annee         1999 non-null   int64         
 2   mois          1999 non-null   int64         
 3   Jour          1999 non-null   datetime64[ns]
 4   mois_nom      1999 non-null   object        
 5   annee_mois    1999 non-null   int64         
 6   jour_semaine  1999 non-null   int64         
 7   trimestre     1999 non-null   object        
dtypes: datetime64[ns](1), int64(5), object(2)
memory usage: 125.1+ KB


Reprend les dates de chaque transaction et les sépares en données atomiques

In [13]:
employes = pd.read_excel(chemin/'Donnees/Copie+de+Extraction+cube+OLAP+-+14+aout+2024.xlsx',sheet_name=4)

In [14]:
employes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56 entries, 0 to 55
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   id_employe  56 non-null     object
 1   employe     56 non-null     object
 2   prenom      56 non-null     object
 3   nom         56 non-null     object
 4   date_debut  56 non-null     int64 
 5   hash_mdp    56 non-null     object
 6   mail        56 non-null     object
dtypes: int64(1), object(6)
memory usage: 3.2+ KB


Cette table contient les identifiants des employés, leur nom prenom abrégés, date de début, le hash de leur mot de passe (accès caisse) et leur mail professionnel. Pour respecter le RGPD on ne garde pas les noms et prénoms complets, ni les hash_mdp.

#### 2.3. Dictionnaire des données

Dictionnaire présent au lien suivant:https://github.com/Sebules/Project_4_OpenClassrooms_Data_Engineer/blob/e8210fe54ad9f05ce24fe45a02a4f1adc3ae3894/Projet4_dico_donnees_Sebastien_Ulesie.xlsx

In [15]:
# Visualisation de chaque sheet du dictionnaire
for sheet in range(0,10):
    dico = pd.read_excel(chemin/'Projet4_dico_donnees_Sebastien_Ulesie.xlsx',sheet_name=sheet)
    display(dico)

,Unnamed: 0,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6
0,DICTIONNAIRE DES DONNÉES - Log Ventes,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,CODE,SIGNIFICATION,TYPE,LONGUEUR,NATURE,REGLE DE GESTION,REGLE DE CALCUL
4,id_user,identifiant de l'utilisateur qui a complété la...,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
5,date,identifiant date de l'opération,Date,NC,Elémentaire,Ne doit pas être nul,NaN
6,action,opération effectuée,Varchar,20,Elémentaire,Ne doit pas être nul,NaN
7,table_insert,table qui a été modifiée. Ici la table Ventes,Varchar,20,Elémentaire,Ne doit pas être nul,NaN
8,id_ligne,identifiant de la vente,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
9,champs,"CUSTUMER_ID, id_employe, EAN, Date, ID_ticket:...",Varchar,20,Elémentaire,Ne doit pas être nul,NaN


,Unnamed: 0,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6
0,DICTIONNAIRE DES DONNÉES - Vente Détail,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,CODE,SIGNIFICATION,TYPE,LONGUEUR,NATURE,REGLE DE GESTION,REGLE DE CALCUL
4,id_bdd,identifiant d'une vente dans la base de données,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
5,customer_id,identifiant du client qui réalise l'achat,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
6,id_employe,identifiant de l'employé (caissier) qui enregi...,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
7,ean,code barre du produit acheté,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
8,date_achat,identifiant date d'achat du produit,Integer,NC,Elémentaire,Ne doit pas être nul,NaN
9,id_ticket,numéro du ticket de caisse,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN


,Unnamed: 0,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6
0,DICTIONNAIRE DES DONNÉES - Log Produits,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,CODE,SIGNIFICATION,TYPE,LONGUEUR,NATURE,REGLE DE GESTION,REGLE DE CALCUL
4,id_user,identifiant de l'utilisateur qui a complété la...,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
5,date,identifiant date de l'opération,Integer,NC,Elémentaire,Ne doit pas être nul,NaN
6,action,opération effectuée,Varchar,20,Elémentaire,Ne doit pas être nul,NaN
7,table_insert,table qui a été modifiée. Ici la table Produits,Varchar,20,Elémentaire,Ne doit pas être nul,NaN
8,id_ligne,identifiant du produit,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
9,champs,prix,Varchar,10,Elémentaire,Ne doit pas être nul,NaN


,Unnamed: 0,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6
0,DICTIONNAIRE DES DONNÉES - Produits,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,CODE,SIGNIFICATION,TYPE,LONGUEUR,NATURE,REGLE DE GESTION,REGLE DE CALCUL
4,ean,code barre du produit acheté,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
5,categorie,"categorie du produit (Boissons, conserves, bou...",Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
6,rayon,rayon du supermarché dans lequel se situe le p...,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
7,libelle_produit,nom du produit,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
8,prix,prix du produit,Float,NC,Elémentaire,Ne doit pas être nul,NaN


,Unnamed: 0,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6
0,DICTIONNAIRE DES DONNÉES - Log Client,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,CODE,SIGNIFICATION,TYPE,LONGUEUR,NATURE,REGLE DE GESTION,REGLE DE CALCUL
4,id_user,identifiant de l'utilisateur qui a complété la...,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
5,date,identifiant date de l'opération,Integer,NC,Elémentaire,Ne doit pas être nul,NaN
6,action,opération effectuée,Varchar,10,Elémentaire,Ne doit pas être nul,NaN
7,table_insert,table qui a été modifiée. Ici la table Client,Varchar,10,Elémentaire,Ne doit pas être nul,NaN
8,id_ligne,identifiant du client,Varchar,50,Elémentaire,Ne doit pas être nul,NaN
9,champs,date_inscription,Varchar,20,Elémentaire,Ne doit pas être nul,NaN


,Unnamed: 0,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6
0,DICTIONNAIRE DES DONNÉES - calendrier,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,CODE,SIGNIFICATION,TYPE,LONGUEUR,NATURE,REGLE DE GESTION,REGLE DE CALCUL
4,date,identifiant date,Integer,NC,Elémentaire,Ne doit pas être nul,NaN
5,annee,annee de la transaction,Integer,4,Elémentaire,Ne doit pas être nul,NaN
6,mois,mois de la transaction,Integer,2,Elémentaire,Ne doit pas être nul,NaN
7,jour,jour de la transaction,Integer,2,Elémentaire,Ne doit pas être nul,NaN
8,mois_nom,nom du mois,Varchar,10,Elémentaire,Ne doit pas être nul,NaN
9,annee_mois,extrait année et mois,Integer,NC,Troncature,Ne doit pas être nul,NaN


,Unnamed: 0,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6
0,DICTIONNAIRE DES DONNÉES - clients,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,CODE,SIGNIFICATION,TYPE,LONGUEUR,NATURE,REGLE DE GESTION,REGLE DE CALCUL
4,customer_id,identifiant du client qui réalise l'achat,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
5,date_inscription,"date d'inscription du client dans la base, cer...",Date,NC,Elémentaire,Ne doit pas être nul,NaN


,Unnamed: 0,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6
0,DICTIONNAIRE DES DONNÉES - Log Employé,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,CODE,SIGNIFICATION,TYPE,LONGUEUR,NATURE,REGLE DE GESTION,REGLE DE CALCUL
4,id_user,identifiant de l'utilisateur qui a complété la...,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
5,date,identifiant date début de l'employé,Integer,NC,Elémentaire,Ne doit pas être nul,NaN
6,action,opération effectuée,Varchar,10,Elémentaire,Ne doit pas être nul,NaN
7,table_insert,table qui a été modifiée. Ici la table Employé,Varchar,10,Elémentaire,Ne doit pas être nul,NaN
8,id_ligne,identifiant de l'employé,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
9,champs,hash_mdp,Varchar,10,Elémentaire,Ne doit pas être nul,NaN


,Unnamed: 0,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6
0,DICTIONNAIRE DES DONNÉES - Employés,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,CODE,SIGNIFICATION,TYPE,LONGUEUR,NATURE,REGLE DE GESTION,REGLE DE CALCUL
4,id_employe,identifiant,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
5,employe,accronyme employé,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
6,prenom,prenom employé,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
7,nom,nom employé,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN
8,date_debut,identifiant date d'entrée dans l'entreprise,Integer,NC,Elémentaire,Ne doit pas être nul,NaN
9,hash_mdp,hash du mot de passe de l'employé,Varchar,NC,Elémentaire,Ne doit pas être nul,NaN


,Unnamed: 0,Unnamed: 1
0,NaN,NaN
1,Dictionnaire des données,Significations
2,CODE,Nom de la ligne dans le fichier Initial
3,SIGNIFICATION,Signification de la colonne
4,TYPE DE VARIABLES,"Varchar, Integer, Date, Float, etc."
5,LONGUEUR,Longueur maximale de la donnée (pas obligatoire)
6,NATURE (E/Ca/CO),"Elémentaire, calculé ou concaténé"
7,REGLE DE GESTION,"Format de la date (jj/mm/aaaa, jj/mm/aa, etc),..."
8,REGLE DE CALCUL,exemple : Colonne A + Colonne B


#### 2.4. Schéma relationnel

Schéma réalisé avec SQL Power Architect. Présent au lien suivant: https://github.com/Sebules/Project_4_OpenClassrooms_Data_Engineer/blob/e8210fe54ad9f05ce24fe45a02a4f1adc3ae3894/SQL_architecture_projet4.pdf

![Schéma relationnel](SQL_architecture_projet4.png)

### Etape 3 - Créez la base de données et des requêtes SQL

#### 3.1. Création de la base de données

In [16]:
#Forcer le chemin exact du .env
load_dotenv(r"C:\Users\sebas\Documents\cours Openclassrooms\Data Engineer\projet 4\.env")

True

In [17]:
# Récupérer le mot de passe PostgreSQL
env = os.getenv("PASSWORD")

In [18]:
password =quote_plus(env)
#le mot de passe est intégré à l'URL. Or,certains caractères ont une signification spéciale dans une URL, comme @,: etc.
## quote_plus() sert à encoder correctement les caractères spéciaux pour qu’ils soient valides dans une URL

In [19]:
# Connexion à ma base PostgreSQL locale
engine = create_engine(f'postgresql://postgres:{password}@localhost:5432/smartmarket')

In [20]:
sql_commands ="""
CREATE TABLE IF NOT EXISTS calendrier (
                date INTEGER NOT NULL,
                annee INTEGER NOT NULL,
                mois INTEGER NOT NULL,
                jour INTEGER NOT NULL,
                mois_nom VARCHAR(10) NOT NULL,
                annee_mois INTEGER NOT NULL,
                jour_semaine INTEGER NOT NULL,
                trimestre VARCHAR(2) NOT NULL,
                CONSTRAINT calendrier_pk PRIMARY KEY (date)
);
COMMENT ON COLUMN calendrier.trimestre IS 'Q1, Q2, Q3, Q4';


CREATE TABLE IF NOT EXISTS employe (
                id_employe VARCHAR NOT NULL,
                employe VARCHAR NOT NULL,
                prenom VARCHAR NOT NULL,
                nom VARCHAR NOT NULL,
                date_debut INTEGER NOT NULL,
                hash_mdp VARCHAR NOT NULL,
                mail VARCHAR(100) NOT NULL,
                CONSTRAINT employe_pk PRIMARY KEY (id_employe)
);


CREATE TABLE IF NOT EXISTS client (
                customer_id VARCHAR NOT NULL,
                date_inscription DATE NOT NULL,
                CONSTRAINT client_pk PRIMARY KEY (customer_id)
);


CREATE TABLE IF NOT EXISTS produit (
                ean VARCHAR NOT NULL,
                categorie VARCHAR NOT NULL,
                rayon VARCHAR NOT NULL,
                libelle_produit VARCHAR NOT NULL,
                prix REAL NOT NULL,
                CONSTRAINT produit_pk PRIMARY KEY (ean)
);


CREATE TABLE IF NOT EXISTS vente (
                id_bdd VARCHAR NOT NULL,
                customer_id VARCHAR NOT NULL,
                id_employe VARCHAR NOT NULL,
                ean VARCHAR NOT NULL,
                date INTEGER NOT NULL,
                id_ticket VARCHAR NOT NULL,
                CONSTRAINT vente_pk PRIMARY KEY (id_bdd)
);
"""
sql_commands = [cmd.strip() for cmd in sql_commands.split(";") if cmd.strip()]
with engine.connect() as conn:
    for cmd in sql_commands:
        conn.execute(text(cmd))
    conn.commit()

In [21]:
# sql_commands = """
# ALTER TABLE vente ADD CONSTRAINT calendrier_vente_fk
# FOREIGN KEY (date_achat)
# REFERENCES calendrier (date_achat)
# ON DELETE NO ACTION
# ON UPDATE NO ACTION
# NOT DEFERRABLE;

# ALTER TABLE vente ADD CONSTRAINT employe_vente_fk
# FOREIGN KEY (id_employe)
# REFERENCES employe (id_employe)
# ON DELETE NO ACTION
# ON UPDATE NO ACTION
# NOT DEFERRABLE;

# ALTER TABLE vente ADD CONSTRAINT client_vente_fk
# FOREIGN KEY (customer_id)
# REFERENCES client (customer_id)
# ON DELETE NO ACTION
# ON UPDATE NO ACTION
# NOT DEFERRABLE;

# ALTER TABLE vente ADD CONSTRAINT produit_vente_fk
# FOREIGN KEY (ean)
# REFERENCES produit (ean)
# ON DELETE NO ACTION
# ON UPDATE NO ACTION
# NOT DEFERRABLE;

# """
# with engine.connect() as conn:
#     for cmd in sql_commands:
#         conn.execute(text(cmd))
#     conn.commit()

#### 3.2 Chargement des données

Lors de l'import du calendrier, bien s'assurer que les dates sont dans le bon format (yyyy-mm-dd) sinon, toutes les lignes ne s'importe pas.

#### 3.3 chiffre d'affaires du 14 août

In [22]:
sql_text="""
SELECT
SUM(p.prix) AS chiffre_affaires
FROM vente v 
JOIN produit p
	ON v.ean = p.ean;
"""

with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat

,chiffre_affaires
0,284242.7


Le montant est très proche de 284243,88. la différence est légère. Vérifions si ce n'est pas un problème d'arrondi.

In [23]:
sql_text= """
SELECT 
  SUM(p.prix) AS CA_brute,
  SUM(ROUND(p.prix::numeric, 2)) AS CA_arrondie
FROM vente v
JOIN produit p
	ON v.ean = p.ean
;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat

,ca_brute,ca_arrondie
0,284242.7,284243.88


On a bien le 14 août 284243,88 € de chiffre d'affaires.

#### 3.4. chiffre d’affaires par client pour le top 10 des clients

In [24]:
sql_text="""
SELECT 
  v.customer_id as client,
  SUM(ROUND(p.prix::numeric, 2)) AS ca_arrondie
FROM vente v
join produit p
	on v.ean = p.ean
group by v.customer_id
order by ca_arrondie desc
limit 10;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat

,client,ca_arrondie
0,CUST-JNSOZSFORR88,846.86
1,CUST-GM6VBAYAB8SF,666.86
2,CUST-L2ST2JHI7K9O,644.18
3,CUST-WU7ZKQJE4L17,608.93
4,CUST-9WM83101QDTI,582.03
5,CUST-ZMAOVX8XYGJY,576.39
6,CUST-3K66CV0OHH7Q,571.44
7,CUST-CG23SXJDRNYR,531.09
8,CUST-D8IOFHVUFX3Y,477.35
9,CUST-IHN1HQRI7PYJ,463.73


#### 3.5 la part de chiffre d’affaires encaissé par employé

Créons une vue pour le chiffre d'affaires par employé

In [38]:
sql_command = """
CREATE or REPLACE VIEW ca_per_employe AS(
SELECT
v.id_employe,
SUM(ROUND(p.prix::numeric, 2)) AS ca_arrondie_par_employe
FROM vente v
JOIN produit p
	ON v.ean = p.ean
GROUP BY v.id_employe
ORDER BY ca_arrondie_par_employe DESC
);
"""

with engine.connect() as conn:
    conn.execute(text(sql_command))
    conn.commit()
    

In [39]:
# affichage de la vue.
sql_text="""
SELECT * FROM ca_per_employe
;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat

,id_employe,ca_arrondie_par_employe
0,f491076a1ff2d873ebea809c11144542,7818.82
1,e01e752175e05f00c8314ccb8da4c418,7736.16
2,8d1001fbad3d2a60ff7530600ed5d55e,6995.14
3,6c1c3292c852c6c593b95cc146b00c0e,6616.46
4,a7ada0770091e838e3dcd45265282820,6483.84
5,2477db17c02f512ebc4b20f01a7edb55,6361.22
6,528c733809cb51a3634befb260b5d243,6133.34
7,1c1d83678cc463b52366dae07cd14c8a,6112.37
8,dd595f0f0b3400df2908f0be7723dad4,6111.18
9,c4f0909120cfde7086a3c1a56e96a015,6094.62


Créons une vue pour avoir le chiffre d'affaires total sur chaque ligne qui représente un employé.

In [42]:
sql_command = """
CREATE or REPLACE VIEW ca_arrondi AS(
SELECT
id_employe,
SUM(ca_arrondie_par_employe) OVER() AS ca_arrondi
FROM ca_per_employe
);
"""
with engine.connect() as conn:
    conn.execute(text(sql_command))
    conn.commit()

In [43]:
# affichage de la vue.
sql_text="""
SELECT * FROM ca_arrondi
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat

,id_employe,ca_arrondi
0,f491076a1ff2d873ebea809c11144542,284243.88
1,e01e752175e05f00c8314ccb8da4c418,284243.88
2,8d1001fbad3d2a60ff7530600ed5d55e,284243.88
3,6c1c3292c852c6c593b95cc146b00c0e,284243.88
4,a7ada0770091e838e3dcd45265282820,284243.88
5,2477db17c02f512ebc4b20f01a7edb55,284243.88
6,528c733809cb51a3634befb260b5d243,284243.88
7,1c1d83678cc463b52366dae07cd14c8a,284243.88
8,dd595f0f0b3400df2908f0be7723dad4,284243.88
9,c4f0909120cfde7086a3c1a56e96a015,284243.88


Affichons la part de chaque employé dans une chiffre d'affaires

In [44]:
sql_text="""
SELECT
e.employe,
CONCAT(c.jour,'/',c.mois,'/',c.annee) AS date_de_début,
cap.ca_arrondie_par_employe,
CONCAT(ROUND(cap.ca_arrondie_par_employe*100.0/ca_arrondi,2),'%') AS part_ca_par_employe
FROM ca_per_employe cap
JOIN ca_arrondi ca ON cap.id_employe = ca.id_employe
JOIN employe e ON e.id_employe = ca.id_employe
JOIN calendrier c ON e.date_debut = c.date
ORDER BY part_ca_par_employe DESC
;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat

,employe,date_de_début,ca_arrondie_par_employe,part_ca_par_employe
0,aboulet,14/11/2023,7818.82,2.75%
1,ejacquier,12/6/2020,7736.16,2.72%
2,cdelisle,2/12/2020,6995.14,2.46%
3,pmanoury,29/9/2020,6616.46,2.33%
4,tarsenault,11/1/2022,6483.84,2.28%
5,adufresne,25/7/2023,6361.22,2.24%
6,adutertre,2/3/2023,6133.34,2.16%
7,pange,2/5/2021,6112.37,2.15%
8,alièvremont,12/6/2020,6111.18,2.15%
9,prodier,20/1/2021,6094.62,2.14%


## Partie 2 - Analysez les logs de la base de données

De : Hugo

À : Moi

Objet : Étude des logs

Hello,

 

Merci pour ton support de présentation avec tous les éléments sur la base de données et sur les résultats de tes requêtes SQL. 

 

Tu as pu confirmer que le chiffre d’affaires est bien de 284 243,88€. Cependant, nous ne comprenons pas pourquoi ce chiffre n’est pas constant. 

 

Pour que tu puisses continuer à investiguer sur ce sujet, tu trouveras en pièces joints les logs de cette semaine d'août.

 

Est-ce que tu peux étudier les logs et nous dire si tu trouves quelque chose ? 

 

N’oublie pas de reporter tes informations dans le support de présentation. 

 

Merci encore pour ton aide précieuse !

 

A très vite !

 

Hugo
Responsable de l’équipe BI

 

PS : je pense que le plus simple c’est d’ajouter ce fichier dans ta base de données car il n’y a que des ID dans les logs. 

### Etape 1 - Insérez les logs dans votre base de données et comprenez-les

In [45]:
logs_1 = pd.read_excel(chemin/'Donnees/Logs.xlsx')

In [46]:
logs = logs_1.copy()

In [47]:
logs.head()

,id_user,date,action,table_insert,id_ligne,champs,detail,Unnamed: 7
0,0859b70a9ca905a870bfccc5f8440ba1,45518,INSERT,Client,CUST-0P9K9MICI4V3,date_inscription,2024-08-14 00:00:00,NaN
1,c4f0909120cfde7086a3c1a56e96a015,45518,INSERT,Client,CUST-SZ9IS3Q1LBEP,date_inscription,2024-08-14 00:00:00,NaN
2,dd595f0f0b3400df2908f0be7723dad4,45518,INSERT,Client,CUST-L99QEXFV9WTS,date_inscription,2024-08-14 00:00:00,NaN
3,e44eabec860498a79e675200bc8fa455,45518,INSERT,Client,CUST-9SBDRZ728025,date_inscription,2024-08-14 00:00:00,NaN
4,37c6a856b2e14217855d808fbc8e56a7,45518,INSERT,Client,CUST-3F47X2KHYXQM,date_inscription,2024-08-14 00:00:00,NaN


Les logs indiquent les actions qui ont lieu sur les tables de la base de données par un utilisateur donnée à une date précise.

In [48]:
logs['table_insert'].value_counts(dropna=False)

table_insert
Ventes      206885
Produits       575
Client          20
Employé          9
Name: count, dtype: int64

4 tables ont été modifiées, ventes, produits, client et employé. Séparons les logs pour avoir les logs pour chacune des tables.

In [49]:
# Mettre la date au bon format
logs["date"] = pd.to_datetime(logs["date"], unit='D', origin='1899-12-30')

In [50]:
logs.head(1)

,id_user,date,action,table_insert,id_ligne,champs,detail,Unnamed: 7
0,0859b70a9ca905a870bfccc5f8440ba1,2024-08-14,INSERT,Client,CUST-0P9K9MICI4V3,date_inscription,2024-08-14 00:00:00,NaN


In [51]:
logs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207489 entries, 0 to 207488
Data columns (total 8 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   id_user       207489 non-null  object        
 1   date          207489 non-null  datetime64[ns]
 2   action        207489 non-null  object        
 3   table_insert  207489 non-null  object        
 4   id_ligne      207489 non-null  object        
 5   champs        207487 non-null  object        
 6   detail        207487 non-null  object        
 7   Unnamed: 7    199990 non-null  float64       
dtypes: datetime64[ns](1), float64(1), object(6)
memory usage: 12.7+ MB


In [52]:
logs['Unnamed: 7'].value_counts(dropna=False)

Unnamed: 7
0.0    199990
NaN      7499
Name: count, dtype: int64

In [53]:
logs = logs.drop('Unnamed: 7',axis=1)

In [54]:
logs_filtre=logs[(logs['table_insert']=='Ventes') & (logs['champs']=='Date')]

In [55]:
logs_filtre.assign(detail= lambda x: pd.to_datetime(x['detail'].astype(float),unit='D', origin='1899-12-30')).head()
#vérification avant de le mettre dans logs.

,id_user,date,action,table_insert,id_ligne,champs,detail
607,08c8b678f8e6f0caz05880ef4ebba10az,2024-08-14,INSERT,Ventes,000MDBR8LMMFI8BWFEYR4KEE8MYGUPJVPEO,Date,2024-08-14
612,08c8b678f8e6f0caz05880ef4ebba10az,2024-08-14,INSERT,Ventes,0031A86HBT87DS7FU1QBUN36UD67WEW417N,Date,2024-08-14
617,08c8b678f8e6f0caz05880ef4ebba10az,2024-08-14,INSERT,Ventes,003RNQDBFXN5XKH5HRES27UE70871U0MLFU,Date,2024-08-14
622,08c8b678f8e6f0caz05880ef4ebba10az,2024-08-14,INSERT,Ventes,004YV4B4FHMRS7VF8UXV8WFOU7LRW06GOCV,Date,2024-08-14
627,08c8b678f8e6f0caz05880ef4ebba10az,2024-08-14,INSERT,Ventes,005THF0N473QYI1ZZQG3PZ8G3135L7LJQKA,Date,2024-08-14


In [56]:
logs.loc[(logs['table_insert']=='Ventes') & (logs['champs']=='Date'),'detail'] = logs.loc[(logs['table_insert']=='Ventes') & (logs['champs']=='Date')].assign(detail= lambda x: pd.to_datetime(x['detail'].astype(float),unit='D', origin='1899-12-30'))

In [57]:
logs[(logs['table_insert']=='Ventes') & (logs['champs']=='Date')].head()

,id_user,date,action,table_insert,id_ligne,champs,detail
607,08c8b678f8e6f0caz05880ef4ebba10az,2024-08-14,INSERT,Ventes,000MDBR8LMMFI8BWFEYR4KEE8MYGUPJVPEO,Date,2024-08-14 00:00:00
612,08c8b678f8e6f0caz05880ef4ebba10az,2024-08-14,INSERT,Ventes,0031A86HBT87DS7FU1QBUN36UD67WEW417N,Date,2024-08-14 00:00:00
617,08c8b678f8e6f0caz05880ef4ebba10az,2024-08-14,INSERT,Ventes,003RNQDBFXN5XKH5HRES27UE70871U0MLFU,Date,2024-08-14 00:00:00
622,08c8b678f8e6f0caz05880ef4ebba10az,2024-08-14,INSERT,Ventes,004YV4B4FHMRS7VF8UXV8WFOU7LRW06GOCV,Date,2024-08-14 00:00:00
627,08c8b678f8e6f0caz05880ef4ebba10az,2024-08-14,INSERT,Ventes,005THF0N473QYI1ZZQG3PZ8G3135L7LJQKA,Date,2024-08-14 00:00:00


Il y a des prix qui ont été enregistrés comme des dates. Du coup, il faut le convertir en float car on aura des valeurs aberrantes lors de l'import. 

In [58]:
mask_date = logs['champs'] == 'prix'
mask_type = logs['detail'].apply(lambda x: isinstance(x, datetime)) # récupérer que les données qui sont des dates

In [59]:
logs_filtre_prix = logs[mask_date & mask_type] # pour récupérer les prix qui sont des dates

In [60]:
logs_filtre_prix.head()

,id_user,date,action,table_insert,id_ligne,champs,detail
21,08c8b678f8e6f0caz05880ef4ebba10az,2024-08-14,UPDATE,Produits,2417137162257,prix,2024-08-02 00:00:00
27,08c8b678f8e6f0caz05880ef4ebba10az,2024-08-14,UPDATE,Produits,5703530830641,prix,2024-02-03 00:00:00
34,08c8b678f8e6f0caz05880ef4ebba10az,2024-08-14,UPDATE,Produits,2038477850150,prix,2024-07-04 00:00:00
35,08c8b678f8e6f0caz05880ef4ebba10az,2024-08-14,UPDATE,Produits,8934396548263,prix,2024-11-03 00:00:00
36,08c8b678f8e6f0caz05880ef4ebba10az,2024-08-14,UPDATE,Produits,9387784092713,prix,2024-08-01 00:00:00


In [61]:
pd.to_datetime(logs_filtre_prix['detail']).dt.day

21      2
27      3
34      4
35      3
36      1
       ..
589     3
591     3
592     6
593    15
594     4
Name: detail, Length: 136, dtype: int32

In [62]:
pd.to_datetime(logs_filtre_prix['detail']).dt.month/100

21     0.08
27     0.02
34     0.07
35     0.11
36     0.08
       ... 
589    0.09
591    0.05
592    0.09
593    0.09
594    0.09
Name: detail, Length: 136, dtype: float64

In [63]:
pd.to_datetime(logs_filtre_prix['detail']).dt.day+pd.to_datetime(logs_filtre_prix['detail']).dt.month/100

21      2.08
27      3.02
34      4.07
35      3.11
36      1.08
       ...  
589     3.09
591     3.05
592     6.09
593    15.09
594     4.09
Name: detail, Length: 136, dtype: float64

In [64]:
combined_mask = mask_date & mask_type

In [65]:
def convert_detail(x):
    if isinstance(x, datetime):
        return float(x.day + x.month / 100)
    return x

In [66]:
logs.loc[combined_mask, 'detail'] = logs.loc[combined_mask, 'detail'].apply(convert_detail)

In [67]:
logs['detail'].apply(lambda x: isinstance(x, datetime))

0          True
1          True
2          True
3          True
4          True
          ...  
207484    False
207485    False
207486    False
207487     True
207488    False
Name: detail, Length: 207489, dtype: bool

In [68]:
logs['table_insert'].unique()

array(['Client', 'Produits', 'Employé', 'Ventes'], dtype=object)

In [69]:
# for table in logs['table_insert'].unique():
#     df = logs[logs['table_insert']==table]
#     table = unidecode(table).lower()
#     df.to_csv(chemin/f'Donnees/logs_{table}.csv',encoding='utf-8')
    

#### 1.2. Création des tables dans PostgreSQL

In [70]:
sql_command = """
CREATE TABLE IF NOT EXISTS logs_client (
id_user VARCHAR NOT NULL,
date INTEGER NOT NULL,
action VARCHAR(10) NOT NULL,
table_insert VARCHAR(10) NOT NULL,
id_ligne VARCHAR(50) NOT NULL,
champs VARCHAR(20) NOT NULL,
detail DATE NOT NULL
);
"""
with engine.connect() as conn:
    conn.execute(text(sql_command))
    conn.commit()

In [71]:
sql_text = """
SELECT * FROM logs_client;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat

,id_user,date,action,table_insert,id_ligne,champs,detail
0,0859b70a9ca905a870bfccc5f8440ba1,45518,INSERT,Client,CUST-0P9K9MICI4V3,date_inscription,2024-08-14
1,c4f0909120cfde7086a3c1a56e96a015,45518,INSERT,Client,CUST-SZ9IS3Q1LBEP,date_inscription,2024-08-14
2,dd595f0f0b3400df2908f0be7723dad4,45518,INSERT,Client,CUST-L99QEXFV9WTS,date_inscription,2024-08-14
3,e44eabec860498a79e675200bc8fa455,45518,INSERT,Client,CUST-9SBDRZ728025,date_inscription,2024-08-14
4,37c6a856b2e14217855d808fbc8e56a7,45518,INSERT,Client,CUST-3F47X2KHYXQM,date_inscription,2024-08-14
5,07c8b678f8e6f0cb04480ef9ebba10ce,45518,INSERT,Client,CUST-DNIO2MRJHQEJ,date_inscription,2024-08-14
6,a7e6deb00de876cafcf64809cd659d0c,45518,INSERT,Client,CUST-XOG109TDC4BT,date_inscription,2024-08-14
7,f491076a1ff2d873ebea809c11144542,45518,INSERT,Client,CUST-ZZ6ASS2HRJ5B,date_inscription,2024-08-14
8,dd595f0f0b3400df2908f0be7723dad4,45518,INSERT,Client,CUST-EXZZR3IHR279,date_inscription,2024-08-14
9,6fa61d0ecae0b563fef18d36b2039c8e,45518,INSERT,Client,CUST-SX94Q0ZUD0LT,date_inscription,2024-08-14


In [72]:
sql_command = """
CREATE TABLE IF NOT EXISTS logs_employe (
id_user VARCHAR NOT NULL,
date INTEGER NOT NULL,
action VARCHAR(10) NOT NULL,
table_insert VARCHAR(10) NOT NULL,
id_ligne VARCHAR NOT NULL,
champs VARCHAR(10),
detail VARCHAR
);
"""
with engine.connect() as conn:
    conn.execute(text(sql_command))
    conn.commit()

In [73]:
sql_text = """
SELECT * FROM logs_employe;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat

,id_user,date,action,table_insert,id_ligne,champs,detail
0,08c8b678f8e6f0caz05880ef4ebba10az,45518,UPDATE,Employé,f6cd8ba3485769b3ad9bab5b7725858e,hash_mdp,a5b1d0e2f3456789cdef1234abcd5678
1,08c8b678f8e6f0caz05880ef4ebba10az,45518,UPDATE,Employé,0859b70a9ca905a870bfccc5f8440ba1,hash_mdp,b2c3d4e5f6789abcde012345f9a8b7c6
2,08c8b678f8e6f0caz05880ef4ebba10az,45518,UPDATE,Employé,07c8b678f8e6f0cb04480ef9ebba10ce,hash_mdp,e9f8d7c6b5a4e3d2c1f098765432abcd
3,08c8b678f8e6f0caz05880ef4ebba10az,45518,UPDATE,Employé,342281771d02d2096972e38e78e7d6bd,hash_mdp,f4e3d2c1b0987a65cdef1234a5b67890
4,08c8b678f8e6f0caz05880ef4ebba10az,45518,UPDATE,Employé,a7ada0770091e838e3dcd45265282820,hash_mdp,c2d3e4f56789a0b1cdef123456789abc
5,08c8b678f8e6f0caz05880ef4ebba10az,45518,UPDATE,Employé,3a0271d540734228c98bcfc2fafe8277,hash_mdp,b7a8c9d0e1f23456789abcd0123456ef
6,08c8b678f8e6f0caz05880ef4ebba10az,45518,UPDATE,Employé,9f29007d63c46217deeb64efc81eba9e,hash_mdp,e2f3a4c5d6b7890ab1234cdef56789a0
7,08c8b678f8e6f0caz05880ef4ebba10az,45518,DELETE,Employé,c3d4e5f6789a0b1c2def3456789abcde1,None,None
8,08c8b678f8e6f0caz05880ef4ebba10az,45518,DELETE,Employé,f6e7d8c9b0a1e2d3f456789abc1234567,None,None


In [74]:
sql_command = """
CREATE TABLE IF NOT EXISTS logs_produits (
id_user VARCHAR NOT NULL,
date INTEGER NOT NULL,
action VARCHAR(10) NOT NULL,
table_insert VARCHAR(10) NOT NULL,
id_ligne VARCHAR NOT NULL,
champs VARCHAR(5) NOT NULL,
detail FLOAT NOT NULL
);
"""
with engine.connect() as conn:
    conn.execute(text(sql_command))
    conn.commit()

In [75]:
sql_text = """
SELECT * FROM logs_produits;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat.head()

,id_user,date,action,table_insert,id_ligne,champs,detail
0,08c8b678f8e6f0caz05880ef4ebba10az,45518,UPDATE,Produits,8855635668141,prix,2.54
1,08c8b678f8e6f0caz05880ef4ebba10az,45518,UPDATE,Produits,2417137162257,prix,2.08
2,08c8b678f8e6f0caz05880ef4ebba10az,45518,UPDATE,Produits,3190630177348,prix,1.89
3,08c8b678f8e6f0caz05880ef4ebba10az,45518,UPDATE,Produits,526744548452,prix,2.49
4,08c8b678f8e6f0caz05880ef4ebba10az,45518,UPDATE,Produits,7411694986259,prix,3.95


In [76]:
sql_command = """
CREATE TABLE IF NOT EXISTS logs_ventes (
id_user VARCHAR NOT NULL,
date INTEGER NOT NULL,
action VARCHAR(10) NOT NULL,
table_insert VARCHAR(10) NOT NULL,
id_ligne VARCHAR NOT NULL,
champs VARCHAR(20) NOT NULL,
detail VARCHAR NOT NULL
);
"""
with engine.connect() as conn:
    conn.execute(text(sql_command))
    conn.commit()

In [77]:
sql_text = """
SELECT * FROM logs_ventes;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat.head()

,id_user,date,action,table_insert,id_ligne,champs,detail
0,08c8b678f8e6f0caz05880ef4ebba10az,45518,INSERT,Ventes,000MDBR8LMMFI8BWFEYR4KEE8MYGUPJVPEO,CUSTUMER_ID,CUST-F6TIGKY2H6WR
1,08c8b678f8e6f0caz05880ef4ebba10az,45518,INSERT,Ventes,000MDBR8LMMFI8BWFEYR4KEE8MYGUPJVPEO,id_employe,4853b03deab973a1a8e466025bce5b52
2,08c8b678f8e6f0caz05880ef4ebba10az,45518,INSERT,Ventes,000MDBR8LMMFI8BWFEYR4KEE8MYGUPJVPEO,EAN,1987784879907
3,08c8b678f8e6f0caz05880ef4ebba10az,45518,INSERT,Ventes,000MDBR8LMMFI8BWFEYR4KEE8MYGUPJVPEO,Date,45518
4,08c8b678f8e6f0caz05880ef4ebba10az,45518,INSERT,Ventes,000MDBR8LMMFI8BWFEYR4KEE8MYGUPJVPEO,ID ticket,t_3839


### Etape 2 - Corrigez dynamiquement la base de données avec les logs

Vérifions si les clients qui sont indiqués comme INSERT ont bien été ajoutés à la base client le 14/08/2024:

In [78]:
sql_text="""
SELECT
lc.id_ligne,
c.date_inscription
FROM logs_client lc
JOIN client c
	ON lc.id_ligne = c.customer_id
;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat

,id_ligne,date_inscription
0,CUST-0P9K9MICI4V3,2024-08-14
1,CUST-SZ9IS3Q1LBEP,2024-08-14
2,CUST-L99QEXFV9WTS,2024-08-14
3,CUST-9SBDRZ728025,2024-08-14
4,CUST-3F47X2KHYXQM,2024-08-14
5,CUST-DNIO2MRJHQEJ,2024-08-14
6,CUST-XOG109TDC4BT,2024-08-14
7,CUST-ZZ6ASS2HRJ5B,2024-08-14
8,CUST-EXZZR3IHR279,2024-08-14
9,CUST-SX94Q0ZUD0LT,2024-08-14


Les 20 nouveaux clients ont bien été ajoutés à la base.

Vérifions si ces clients ont fait des achats le 14 août.

In [84]:
sql_text="""
SELECT
lc.id_ligne AS client,
v.date,
CONCAT(c.jour,'/',c.mois,'/',c.annee) AS date_insertion,
v.id_ticket 
FROM logs_client lc
JOIN vente v ON lc.id_ligne = v.customer_id
JOIN calendrier c ON v.date = c.date
GROUP BY client,v.date,v.id_ticket, date_insertion
;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat

,client,date,date_insertion,id_ticket
0,CUST-Z0WB0J6AYKTM,45518,14/8/2024,t_2670
1,CUST-ED1KQUW44BQX,45518,14/8/2024,t_4641
2,CUST-99B8Q0SVYXJ7,45518,14/8/2024,t_1361
3,CUST-EXZZR3IHR279,45518,14/8/2024,t_7
4,CUST-SZ9IS3Q1LBEP,45518,14/8/2024,t_13
5,CUST-X72AW6JN07ZS,45518,14/8/2024,t_711
6,CUST-WGFE8J11VQFE,45518,14/8/2024,t_1324
7,CUST-0P9K9MICI4V3,45518,14/8/2024,t_2262
8,CUST-9BJVABA8K09C,45518,14/8/2024,t_4453
9,CUST-Y9MQQBQ6YSFJ,45518,14/8/2024,t_617


Les 20 nouveaux clients ajoutés, ont bien fait des achats ce jour-là. 

Attardons-nous sur les produits. L'analyse du log produit avait montré que certains prix avaient été enregistrés en date. Vérifions si leur mise à jour a bien été prise en compte. 

In [86]:
sql_text="""
SELECT
lp.id_ligne AS code_barre,
p.libelle_produit,
p.prix AS prix_bdd,
lp.detail AS prix_log,
ROUND((p.prix-lp.detail)::numeric,2) AS difference
FROM logs_produits lp
JOIN produit p ON lp.id_ligne = p.ean
ORDER BY difference DESC
;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat.head()


,code_barre,libelle_produit,prix_bdd,prix_log,difference
0,8855635668141,500g nids tricolores gd mere,2.54,2.54,0.0
1,2417137162257,450g riz complet lustucru,2.08,2.08,0.0
2,3190630177348,500g semoule moyenne alpina,1.89,1.89,0.0
3,526744548452,100g pralin vahine,2.49,2.49,0.0
4,7411694986259,100g citrons confits,3.95,3.95,0.0


In [90]:
sql_text="""
SELECT
COUNT(*)
FROM logs_produits lp
JOIN produit p ON lp.id_ligne = p.ean
;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat.head()


,count
0,575


Les 575 produits dont les prix ont été mis à jour sont bien présents dans la base produit

Vérifions si toutes les ventes ont bien été ajoutées à la base de données.

In [96]:
sql_text="""
SELECT
lv.id_ligne AS vente,
v.id_ticket,
lv.date AS date_log
FROM logs_ventes lv
JOIN vente v
	ON lv.id_ligne = v.id_bdd
GROUP BY lv.id_ligne,v.id_ticket,lv.date
;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat.head()

,vente,id_ticket,date_log
0,H9KJAXTXQP5A6LL0JW2QT6TPJT984AFFZX2,t_3590,45518
1,F585ZCD4W7ESE2E9KHDXF1MYDYM3DLEMUKB,t_4582,45518
2,U1QKUPOJQ1HMZ86537Y40U17KZ013RDQN6U,t_1252,45518
3,HJKRMNY7KB4UVN2JCY9XN01OJE0C8NQLSOA,t_3413,45518
4,L682TC7HBIOL8ZF7UWJPL0Q4F24NZRICNME,t_1883,45518


In [97]:
resultat.shape[0]

41377

In [104]:
sql_text="""
WITH logs_ventes_join AS (
SELECT
lv.id_ligne AS vente,
v.id_ticket,
lv.date AS date_log
FROM logs_ventes lv
JOIN vente v
	ON lv.id_ligne = v.id_bdd
GROUP BY lv.id_ligne,v.id_ticket,lv.date
)
SELECT
COUNT(*)
FROM logs_ventes_join
;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat

,count
0,41377


In [105]:
sql_text= """
-- Vérifions les dates insertion des données de ventes qui sont indiqués dans le log
SELECT 
lv.table_insert,
CONCAT(c.jour,'/',c.mois,'/',c.annee) AS date_insertion
FROM 
logs_ventes lv
JOIN calendrier c
	ON lv.date = c.date
GROUP BY date_insertion,lv.table_insert
;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat.head()

,table_insert,date_insertion
0,Ventes,14/8/2024
1,Ventes,15/8/2024


L'insertion de certaines ventes s'est faite le 15-08-2024 d'après le log.

Vérifions si la différence de CA ne correspond pas à la somme des ventes qui ont été insérées le 15 août

In [106]:
sql_text="""
WITH ventes_log_15 AS (
SELECT
lv.id_ligne AS vente,
CONCAT(c.jour,'/',c.mois,'/',c.annee) AS date_log,
v.ean
FROM logs_ventes lv
JOIN vente v ON lv.id_ligne = v.id_bdd
JOIN calendrier c ON lv.date = c.date
WHERE lv.date = '45519'
GROUP BY lv.id_ligne,v.ean, date_log
)

SELECT
vl.date_log,
ROUND(SUM(p.prix::numeric),2) AS CA_update_15_aout
FROM ventes_log_15 vl
JOIN produit p
ON vl.ean = p.ean
GROUP BY vl.date_log
;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat

,date_log,ca_update_15_aout
0,15/8/2024,9057.29


Euréka! La différence entre 275 186,59€ et 284 243,88€ est bien 9057,29€. Il semble donc que l'insertion de certaines ventes se fait le jour suivant.

In [107]:
sql_text="""
-- Vérifions si les dates des ventes sont bien au 14 août pour les dates de log au 15 août.

WITH logs_ventes_date AS (
SELECT
lv.date AS date_log,
lv.table_insert,
lv.champs,
lv.detail
FROM logs_ventes lv
WHERE lv.champs = 'Date'
GROUP BY date_log, lv.champs, lv.detail,lv.table_insert
)
SELECT
lvd.date_log,
lvd.table_insert,
lvd.champs,
lvd.detail,
CONCAT(c.jour,'/',c.mois,'/',c.annee) AS date_ventes
FROM logs_ventes_date lvd
JOIN calendrier c ON lvd.detail::numeric = c.date
;
"""
with engine.connect() as conn:
    resultat=pd.read_sql(text(sql_text), conn)
resultat

,date_log,table_insert,champs,detail,date_ventes
0,45519,Ventes,Date,45518,14/8/2024
1,45518,Ventes,Date,45518,14/8/2024


## Partie 3 - Faites des recommandations pour l'entreprise

De : Hugo

À : Moi

Objet : Présentation à l’équipe Data

Re,


Comme discuté tout à l’heure à la machine à café, j'aimerais capitaliser sur le travail que tu as déjà réalisé !


Afin de comprendre et de pouvoir documenter l’ensemble de ton travail : 

premièrement, peux-tu rédiger un rapport d’audit sur tout ce que tu as vu ? Tu trouveras en pièce jointe un ancien rapport à adapter à ta situation (je ne t’ai laissé que le sommaire).
deuxièmement, peux-tu également renseigner ton cheminement pour résoudre le problème et les résultats dans ton support de présentation et venir nous présenter ton travail lors d’une réunion d’équipe ? Cela va sensibiliser tout le monde sur la nécessité d’améliorer la qualité et la compréhension de nos données. 
troisièmement, est-ce que tu peux intégrer à ton prototype les différentes mesures correctives que tu auras rédigées dans ton rapport d’audit afin qu’on comprenne les impacts sur notre base de données OLAP ? J’ai entendu parler de différentes choses comme les triggers, les contraintes, les clés étrangères, les propriétés ACID, les procédures stockées, l’analyse des logs systématiques, etc. 
enfin, peux-tu intégrer ton prototype dans le support de présentation ? 

Je te laisse réfléchir et me faire des propositions afin que nos données gagnent en résilience. 


Encore merci pour tout ton travail ! Cela ne fait que quelques semaines mais tu nous aides déjà énormément. 


Hugo
Responsable de l’équipe BI